File: sdk-node/lib/session/ProctorSession.js
const EventEmitter = require('events');

/**
 * Represents a single, active proctoring session for one student.
 * It emits events that the developer can listen to.
 *
 * @fires ProctorSession#violation
 * @fires ProctorSession#event
 * @fires ProctorSession#disconnect
 */
class ProctorSession extends EventEmitter {
    /**
     * @param {string} sessionId The unique ID for this session.
     * @param {WebSocket} webSocket The underlying WebSocket connection.
     */
    constructor(sessionId, webSocket) {
        super();
        this._sessionId = sessionId;
        this._webSocket = webSocket;
        this._startTime = new Date();
    }

    /**
     * The unique identifier for this session.
     * @returns {string}
     */
    get sessionId() {
        return this._sessionId;
    }

    /**
     * The time this session was started.
     * @returns {Date}
     */
    get startTime() {
        return this._startTime;
    }

    /**
     * The underlying WebSocket connection object.
     * @returns {WebSocket}
     */
    get connection() {
        return this._webSocket;
    }

    /**
     * Internal method to handle an incoming event from the client.
     * It parses the event and emits it for the developer to handle.
     * @param {string} eventJson The raw JSON string from the client.
     * @private
     */
    _handleEvent(eventJson) {
        try {
            const event = JSON.parse(eventJson);
            
            // Emit a generic 'event' for all messages.
            this.emit('event', event);

            // Emit a specific 'violation' event if the status is 'violation'.
            if (event.status === 'violation') {
                this.emit('violation', event);
            }
        } catch (error) {
            console.error(`[Session ${this.sessionId}] Error parsing event JSON:`, error);
        }
    }

    /**
     * Internal method to signal that the client has disconnected.
     * @private
     */
    _handleDisconnect() {
        this.emit('disconnect');
    }
}

module.exports = ProctorSession;

File: sdk-java/src/main/java/com/proctor/sdk/session/ProctorSession.java
package com.proctor.sdk.session;

import org.java_websocket.WebSocket;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

/**
 * Represents a single, active proctoring session for one student.
 * This class acts as an event emitter, allowing developers to listen for
 * specific events from an individual student's client.
 */
public class ProctorSession {

    private final String sessionId;
    private final WebSocket connection;

    // Listeners for various event types
    private final List<Consumer<JSONObject>> activityListeners = new ArrayList<>();
    private final List<Consumer<JSONObject>> violationListeners = new ArrayList<>();
    private final List<Runnable> disconnectListeners = new ArrayList<>();

    public ProctorSession(String sessionId, WebSocket connection) {
        this.sessionId = sessionId;
        this.connection = connection;
    }

    public String getSessionId() {
        return sessionId;
    }
    
    /**
     * Adds a listener that will be called for any event from the client.
     * @param listener The callback to execute.
     */
    public void onActivity(Consumer<JSONObject> listener) {
        activityListeners.add(listener);
    }

    /**
     * Adds a listener that will be called only for events marked as 'violation'.
     * @param listener The callback to execute.
     */
    public void onViolation(Consumer<JSONObject> listener) {
        violationListeners.add(listener);
    }
    
    /**
     * Adds a listener that will be called when the student's client disconnects.
     * @param listener The callback to execute.
     */
    public void onDisconnect(Runnable listener) {
        disconnectListeners.add(listener);
    }

    /**
     * Called by the server when a message is received for this session.
     * It parses the event and notifies the appropriate listeners.
     * @param message The raw JSON string from the client.
     */
    public void handleMessage(String message) {
        try {
            JSONObject event = new JSONObject(message);
            
            // Notify all general activity listeners
            activityListeners.forEach(listener -> listener.accept(event));
            
            // If the event is a violation, notify the specific violation listeners
            if ("violation".equals(event.optString("status"))) {
                violationListeners.forEach(listener -> listener.accept(event));
            }

        } catch (Exception e) {
            System.err.println("Error parsing event JSON for session " + sessionId + ": " + message);
        }
    }

    /**
     * Called by the server when the connection for this session is closed.
     */
    public void handleDisconnect() {
        disconnectListeners.forEach(Runnable::run);
    }
    
    /**
     * Closes the connection to the student's client.
     */
    public void disconnect() {
        if (connection != null && connection.isOpen()) {
            connection.close();
        }
    }
}
